### PR TITLE
NNS1-3274 display ids in popover menu

### PR DESCRIPTION
**Pull request by coskucinkilic**
# Motivation

Improve the visibility of the account and principal ids by adding them to popover menu.

# Changes

 New AccountDetailsPo has been created
 User’s main ICP account ID and principal ID added to the profile popover, including copy buttons next to each
 TooltipIcon is extended to accept color property for icon color
 TooltipIcon's default icon color has been changed to var(--elements-icons) from var(--tertiary)
 
Final profile popover look:
Screenshot 2024-09-12 at 14 23 50

# Tests

new tests are added for pricipal id, account id and tooltip text

# Todos
- [x]  Add entry to changelog

### CODE REVIEW

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts
Comment (dskloetd) [inline frontend/src/tests/lib/components/header/AccountMenu.spec.ts:1]: Can you also make a test file for `AccountDetails.svelte` which tests the functionality for account details separately?

Then in this file you can have just a single test which makes sure the account details component is present.

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts [hunk 1]
```diff
@@ -84,5 +97,56 @@ describe("AccountMenu", () => {
         expect(() => renderResult.getByRole("menu")).toThrow()
       );
     });
+
+    it("should display main ICP account ID if available", async () => {
+      setAccountsForTesting({
+        main: mockMainAccount,
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/header/AccountMenu.spec.ts:103]: Please set an explicit value for the account identifier in the test, so that it's clear that the account identifier that's being used is the one you expect and it doesn't come from somewhere else.
```
  const accountIdForTesting = "82376428374628347628347263847263847623";
  setAccountsForTesting({
    main: {
      ...mockMainAccount,
      identifier: accountIdForTesting,
    }
  });
```

File: frontend/src/tests/page-objects/AccountDetails.page-object.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    );
+  }
+
+  async getMainIcpAccountIdTooltipId(): Promise<string> {
+    const tooltipIconPo = this.getMainIcpAccountIdTooltipIconPo();
+    return tooltipIconPo.getTooltipPo().getTooltipId();
+  }
+
+  async getMainIcpAccountId(): Promise<string> {
+    return this.getMainIcpAccountIdPo().getDisplayedText();
```
Comment (dskloetd) [inline frontend/src/tests/page-objects/AccountDetails.page-object.ts:35]: If you use `getFullText()` instead of `getDisplayedText()` you can compare it to the actual ID.

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts [hunk 3]
```diff
@@ -84,5 +97,56 @@ describe("AccountMenu", () => {
...
+      });
+
+      const { po: accountDetailsPo, renderResult } = renderComponent();
+
+      await show(renderResult);
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/header/AccountMenu.spec.ts:108]: I've never seen or used this. What does this do or why is it needed?

I always only return the page object from the `renderComponent` function.

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts [hunk 4]
```diff
@@ -84,5 +97,56 @@ describe("AccountMenu", () => {
...
+
+      const mainAccountId = await accountDetailsPo.getMainIcpAccountId();
+
+      if (mainAccountId) {
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/header/AccountMenu.spec.ts:112]: Is this necessary? The expect on the line below would also fail if `mainAccountId` is falsy, right?

File: frontend/src/tests/page-objects/AccountDetails.page-object.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  async getPrincipalId(): Promise<string> {
+    return this.getPrincipalIdPo().getDisplayedText();
```
Comment (dskloetd) [inline frontend/src/tests/page-objects/AccountDetails.page-object.ts:39]: Same: use `getFullText()`.

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts [hunk 6]
```diff
@@ -84,5 +97,56 @@ describe("AccountMenu", () => {
...
+    });
+
+    it("should display principal ID if available", async () => {
+      const { po: accountDetailsPo, renderResult } = renderComponent();
+
+      await show(renderResult);
+
+      const principalId = await accountDetailsPo.getPrincipalId();
+
+      if (principalId) {
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/header/AccountMenu.spec.ts:126]: Same: is the `if` necessary?

I don't think it's good practice to have `if`s in tests. The test should know exactly what state it expects and it should just fail if it's not in the expected state.

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts [hunk 7]
```diff
@@ -8,6 +12,15 @@ describe("AccountMenu", () => {
     await waitFor(() => expect(getByRole("menu")).not.toBeNull());
   };
 
+  const renderComponent = () => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/header/AccountMenu.spec.ts:15]: I would expect `renderComponent` in `Foo.spec.ts` to return the page object of the `Foo` component.
Not of some subcomponent.
`AccountMenu` already has a page object: https://github.com/dfinity/nns-dapp/blob/main/frontend/src/tests/page-objects/AccountMenu.page-object.ts

So you can just add `getAccountDetailsPo()` to it.

File: frontend/src/lib/components/header/AccountDetails.svelte [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  let principalId: string | undefined;
+  let accountId: string | undefined;
+
+  $: principalId = $authStore.identity?.getPrincipal().toText();
+  $: accountId = $icpAccountsStore.main?.identifier;
+</script>
+
+{#if accountId || principalId}
+  <div class="account-details" data-tid="account-details-container">
+    {#if accountId}
```
Comment (mstrasinskis) [inline frontend/src/lib/components/header/AccountDetails.svelte:17]: Although this works, we typically use `isNullish`/`nonNullish` utils for this. 
@dskloetd, what do you think about simplifying this rule?

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts (hunk 3 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/header/AccountMenu.spec.ts:108]: I need to have access to container to use the show function to open the popup menu.

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts (hunk 4 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/header/AccountMenu.spec.ts:112]: Fixed, i believe it was failing due to wrong selectors used and wrote this to fix it but then realised that the pageobject is wrongly configured forgot to delete these if statements.

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts (hunk 6 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/header/AccountMenu.spec.ts:126]: fixed

File: frontend/src/lib/components/header/AccountDetails.svelte [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+  let principalId: string | undefined;
+  let accountId: string | undefined;
+
+  $: principalId = $authStore.identity?.getPrincipal().toText();
+  $: accountId = $icpAccountsStore.main?.identifier;
+</script>
+
+{#if accountId || principalId}
+  <div class="account-details" data-tid="account-details-container">
```
Comment (mstrasinskis) [inline frontend/src/lib/components/header/AccountDetails.svelte:16]: It’s always nice to have a component root in tests, regardless of the result of the if statements. I would suggest wrapping the whole component with `<TestIdWrapper>`. Also, we typically use the `-component` suffix for root-level IDs.

File: frontend/src/lib/components/header/AccountDetails.svelte (hunk 8 above)
Comment (peterpeterparker) [reply frontend/src/lib/components/header/AccountDetails.svelte:17]: @mstrasinskis  I just randomly noticing your comment, so thought about sharing: @AntonioVentilii-DFINITY is working on an eslint rule ([PR](https://github.com/dfinity/oisy-wallet/pull/1486)) to enforce their usage for Oisy. It's not finished but, once done, we can ping you if you want to enjoy the same rule.

Comment (coskucinkilic) [reply frontend/src/lib/components/header/AccountDetails.svelte:17]: I think it is better to keep the utils, fixed now

Comment (mstrasinskis) [reply frontend/src/lib/components/header/AccountDetails.svelte:17]: @peterpeterparker, looks interesting. Thank you for the heads-up.

Comment (mstrasinskis) [reply frontend/src/lib/components/header/AccountDetails.svelte:17]: @coskucinkilic, thanks 👍

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/header/AccountMenu.spec.ts:1]: moved the tests, needs review

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts (hunk 1 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/header/AccountMenu.spec.ts:103]: Use this one thank you for clarificatioin

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts (hunk 3 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/header/AccountMenu.spec.ts:108]: removed now

File: frontend/src/tests/page-objects/AccountDetails.page-object.ts (hunk 2 above)
Comment (coskucinkilic) [reply frontend/src/tests/page-objects/AccountDetails.page-object.ts:35]: done

File: frontend/src/tests/page-objects/AccountDetails.page-object.ts (hunk 5 above)
Comment (coskucinkilic) [reply frontend/src/tests/page-objects/AccountDetails.page-object.ts:39]: done

File: frontend/src/tests/page-objects/AccountMenu.page-object.ts [hunk 10]
```diff
@@ -17,7 +17,7 @@ export class AccountMenuPo extends BasePageObject {
     return this.click("logout");
   }
 
-  getAccountDetailsPo(): AccountDetailsPo {
+  getAccountDetailsPoRoot(): AccountDetailsPo {
```
Comment (dskloetd) [inline frontend/src/tests/page-objects/AccountMenu.page-object.ts:20]: Why did you add `Root` at the end of the method name?

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts [hunk 11]
```diff
@@ -84,5 +86,19 @@ describe("AccountMenu", () => {
...
     });
+
+    it("should render account details component", async () => {
+      const renderResult = render(AccountMenu);
+
+      const accountMenuPo = AccountMenuPo.under(
+        new JestPageObjectElement(renderResult.container)
+      );
+
+      await show(renderResult);
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/header/AccountMenu.spec.ts:97]: Does it work to call `await accountMenuPo.openMenu()` instead?

Sorry, I know you're just following the surrounding code but this test file is unfortunately from before we had page objects.
If you want, would also be great to migrate the whole file to using page objects (in a separate PR).

**Review by dskloetd — COMMENTED** [review]
Almost there!

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts (hunk 11 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/components/header/AccountMenu.spec.ts:97]: i will do it on a separate PR. Do i create tickets for these kind of tasks, or talk with Andrew and he manages everything?

File: frontend/src/tests/page-objects/AccountMenu.page-object.ts (hunk 10 above)
Comment (coskucinkilic) [reply frontend/src/tests/page-objects/AccountMenu.page-object.ts]: I was directly accessing root then I have changed it looking at other files, forgot to change the name. It is fixed now.

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts (hunk 11 above)
Comment (dskloetd) [reply frontend/src/tests/lib/components/header/AccountMenu.spec.ts:97]: Sorry for the other PR I meant changing the rest of the test file to use page objects.

This new test already uses a page object so it should properly use the page object and not use a mixture of methods. So can you still try `await accountMenuPo.openMenu()` and not use the `renderResult` here?

I personally don't make tickets nor ask permission for simple cleanups like that. I usually change a test to use page objects before I start changing the test because I find it easier to work with.

Comment (coskucinkilic) [reply frontend/src/tests/lib/components/header/AccountMenu.spec.ts]: There are lots of things moving in this file and in the upcoming prs due to moving themeSwitcher and SourceCodeButton as well. It is a nice idea to keep everything updated to pageObject before starting, will try to keep that in mind for the next time.

File: frontend/src/lib/components/header/AccountDetails.svelte [hunk 12]
```diff
@@ -42,6 +44,9 @@
 </TestIdWrapper>
 
 <style lang="scss">
+  .text-sm {
```
Comment (dskloetd) [inline frontend/src/lib/components/header/AccountDetails.svelte:47]: Is `sm` an abbreviation for small? It's not really obvious. Can you just write `text-small`?

File: frontend/src/tests/lib/components/header/AccountMenu.spec.ts [hunk 13]
```diff
@@ -84,5 +86,19 @@ describe("AccountMenu", () => {
...
+    it("should render account details component", async () => {
+      const renderResult = render(AccountMenu);
+
+      const accountMenuPo = AccountMenuPo.under(
+        new JestPageObjectElement(renderResult.container)
+      );
+
+      await accountMenuPo.openMenu();
+
+      expect(await accountMenuPo.getAccountDetailsPo().root.isPresent()).toBe(
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/header/AccountMenu.spec.ts:99]: You don't need `.root.`. The `BasePageObject` has a `.isPresent()` method as well.
In general, the tests shouldn't be accessing the elements but only interact through the page objects.

**Review by dskloetd — APPROVED** [review]
Thanks!
